### PR TITLE
It/keras3 pytorch

### PR DESCRIPTION
**Pull request by IgorTatarnikov**
Before submitting a pull request (PR), please read the [contributing guide](https://github.com/brainglobe/.github/blob/main/CONTRIBUTING.md).

Please fill out as much of this template as you can, but if you have any problems or questions, just leave a comment and we will help out :)

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
Migrating away from tensorflow backend.

**What does this PR do?**
Makes torch the default backend, CI tests will only run with a torch backend on python 3.9, 3.10 and 3.11. 

## References

Please reference any existing issues/PRs that relate to this PR.

## How has this PR been tested?
Tested locally and on CI.

## Is this a breaking change?
Hopefully not? If people want to continue using a tensorflow installation all they have to do is run `export KERAS_BACKEND="tensorflow"` and make sure `tensorflow>=2.16.1` 

## Does this PR require an update to the documentation?
Not this one. The PR from `cellfinder-to-keras-3` to `main` will definitely need extra documentation. Might be worth timing the release of the next version with a blog post explaining the changes.

## Checklist:

- [x] The code has been tested locally
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by IgorTatarnikov** [comment]
Numbers from the paper, original cellfinder with a tensorflow backend, default model (lighter points) vs retrained model (darker points)
cellfinder_paper_numbers

`cellfinder`, torch backend, default model (lighter points) vs retrained model (darker points)
cellfinder_torch_backend

Both graphs import the same retrained model from the paper.

**Review by sfmig — APPROVED** [review]
Looking good @IgorTatarnikov ! Mainly minor comments or questions in the review. Some higher-level comments below.

* I know you say in the description that docs are not required atm, but maybe a quick edit to the README is useful so that we don't forget when we return to this. Could be just stating the [need to install a backend](https://github.com/keras-team/keras/#install-with-pip), the need to configure the backend (via env variable for example) if developing, how to run tests locally with all backends etc.
* I ran the tests locally using `tox` with the three backends and got some errors for `torch` that seem to relate to Apple chips. I wrote some more details and a possible workaround in the review.
* The comparison with the previous performance looks great! Not the focus of this review, but I was thinking for the blogpost it would be more relevant to compare the `torch` implementation and the previous TF implementation more explicitly. Right now the comparison is more like: given an implementation, how does default model compare to retrained model.

File: .github/workflows/test_include_guard.yaml [hunk 1]
```diff
@@ -24,11 +24,13 @@ jobs:
         with:
           python-version: '3.10'
```
[the remark this answers is not in the document]
Comment (adamltyson) [reply .github/workflows/test_include_guard.yaml:25]: Definately. That could be saved for a seperate PR though, as we also need to update the rest of BG.

File: pyproject.toml [hunk 2]
```diff
@@ -61,14 +61,14 @@ napari = [
 ]
 # Keras backends
 tf = [
-    "tf-nightly==2.16.0.dev20240101", # pinning to same TF as Keras 3.0
+    "tensorflow>=2.16.1",
 ]
 jax = [
     "jax==0.4.20",
```
[the remark this answers is not in the document]
Comment (adamltyson) [reply pyproject.toml:67]: Yeah, want to avoid such tight specification if we can.

File: pyproject.toml [hunk 3]
```diff
@@ -120,13 +120,14 @@ markers = ["slow: marks tests as slow (deselect with '-m \"not slow\"')"]
...
+envlist = py{39,310,311}-{torch}
 isolated_build = true
 
 [gh-actions]
 python =
-    3.9: py39-{tf,jax}  # On GA python=3.9 job, run tox with the tf and jax environments
-    3.10: py310-{tf,jax}  # On GA python=3.10 job, run tox with the tf and jax environments
+    3.9: py39-{torch}  # On GA python=3.9 job, run tox with the torch environment
+    3.10: py310-{torch}  # On GA python=3.10 job, run tox with the torch environment
+    3.11: py311-{torch}  # On GA python=3.11 job, run tox with the torch environment
```
[the remark this answers is not in the document]
Comment (adamltyson) [reply pyproject.toml:130]: I'm fine with only supporting torch. If we aim to support the others, they should be tested in CI. Otherwise they'll just get abandoned later.

File: pyproject.toml [hunk 4]
```diff
@@ -61,14 +61,14 @@ napari = [
...
 # Keras backends
 tf = [
-    "tf-nightly==2.16.0.dev20240101", # pinning to same TF as Keras 3.0
+    "tensorflow>=2.16.1",
 ]
 jax = [
     "jax==0.4.20",
     "jaxlib==0.4.20"
 ]
 torch = [
```
[the remark this answers is not in the document]
Comment (adamltyson) [reply pyproject.toml:70]: Going through these comments, should we just support torch? I don't see any convincing reason to support the others other than "it would be nice".

File: cellfinder/__init__.py [hunk 5]
```diff
@@ -27,25 +27,22 @@
 # If no backend is configured and installed for Keras, tools cannot be used
 # Check backend is configured
 if not os.getenv("KERAS_BACKEND"):
-    os.environ["KERAS_BACKEND"] = "tensorflow"
-    warnings.warn(
-        "Keras backend not configured, automatically set to Tensorflow"
-    )
+    os.environ["KERAS_BACKEND"] = "torch"
+    warnings.warn("Keras backend not configured, automatically set to Torch")
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply cellfinder/__init__.py:31]: I'll go through and make sure they're all lower case. I have no strong opinions but it'd be good to pick one and stick with it!

File: cellfinder/core/classify/classify.py [hunk 6]
```diff
@@ -48,9 +48,7 @@ def main(
         callbacks = None
 
     # Too many workers doesn't increase speed, and uses huge amounts of RAM
-    workers = get_num_processes(
-        min_free_cpu_cores=n_free_cpus, n_max_processes=max_workers
-    )
+    workers = get_num_processes(min_free_cpu_cores=n_free_cpus)
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply cellfinder/core/classify/classify.py:51]: From what I could tell `max_workers` wasn't passed when the this function was called in `core.py` this meant the default value of 3 was used drastically limiting the number of cores. 

We can either remove `max_workers` completely, set the default value to be large (>64?) or expose it in the brainmapper CLI and the napari plugin.

File: cellfinder/core/classify/resnet.py [hunk 7]
```diff
@@ -133,6 +134,10 @@ def non_residual_block(
     )(x)
     x = BatchNormalization(axis=axis, epsilon=bn_epsilon, name="conv1_bn")(x)
     x = Activation(activation, name="conv1_activation")(x)
+
+    if keras.config.backend() == "torch":
+        pooling_padding = "valid"
```
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply cellfinder/core/classify/resnet.py:139]: I'm not the biggest fan of this because then it always overwrites the `pooling_padding` function parameter, whereas we only care to change it if the backend is torch and the padding is set to `same`. However, if we're just supporting torch, perhaps the default function parameter can just be changed to `valid`?

File: .github/workflows/test_include_guard.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply .github/workflows/test_include_guard.yaml:25]: My vote is saving it for a separate PR after this is merged into main

File: cellfinder/core/classify/classify.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamltyson) [reply cellfinder/core/classify/classify.py:51]: This was originally in there because using lots of workers:
- Used huge amounts of RAM
- Didn't speed things up much (presumably the bottleneck was elsewhere)

Do you know if this is still true?

[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply cellfinder/core/classify/classify.py:51]: I usually used 31 workers on a 32 core CPU when testing the accuracy/performance and didn't notice any excess RAM usage. I'm not sure about the performance, it would be good to profile different number of workers across a small and a large dataset.

File: pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (IgorTatarnikov) [reply pyproject.toml:130]: > Unless the plan is to support only `torch` backend 😅

I'm okay with just supporting a `torch` backend! I'll clear out any reference to other backends and have torch be installed as a dependency.

File: .github/workflows/test_and_deploy.yml [hunk 8]
```diff
@@ -39,6 +39,7 @@ jobs:
     runs-on: ${{ matrix.os }}
     env:
       PYTORCH_ENABLE_MPS_FALLBACK: "1"
+      PYTORCH_MPS_HIGH_WATERMARK_RATIO: "0.0"
```
Comment (adamltyson) [inline .github/workflows/test_and_deploy.yml:42]: Could you comment on what these are, and why we're setting them?
